### ISSUE TITLE
[Feature] --immutable should always apply to manifests, like it does with the lockfile

### ISSUE THREAD

**Issue by isker** [yarnpkg/berry#2712]
- [x] I'd be willing to implement this feature ([contributing guide](https://yarnpkg.com/advanced/contributing))
- [x] This feature is important to have in this repository; a contrib plugin wouldn't do

**Describe the user story**
 
`--immutable` allows users to assert that a `yarn install` was not side-effecting on the lockfile.  But `yarn install` can also be side-effecting on the manifest(s).

For example, `--immutable` ought to be used in CI to validate a developer's change.  If that change includes manifests that are not properly formatted, CI should fail so that other developers do not "spontaneously" get local changes upon running `yarn install`.

**Describe the solution you'd like**

`--immutable` causes `yarn install` to error when manifests are altered, just as it does the lockfile.

**Describe the drawbacks of your solution**

This will cause failures of `yarn install` where they would not have previously happened, and so is a compatibility break.

That being said, I think that there are semantically no drawbacks.  I cannot think of a situation where users would not want their `yarn.lock` to change but would want their `package.json` to change.

Perhaps a user might want to validate, during development, that the lockfile in particular is canonical, but does not care about the manifests?  But this is a pretty contrived example in my mind.  `--immutable` is generally intended for automated environments.

**Describe alternatives you've considered**

`immutablePatterns: [package.json, **/package.json]`.  I think this is only preferable if maintaining total compatibility is a hard requirement.

**Comment by isker** [comment]
@merceyz said on Discord:

> I would be fine with making Manifests (package.json) immutable under --immutable but not in immutablePatterns rather hardcode it in Manifest.persistManifest

**Comment by arcanis** [comment]
I like the idea, but I'm not a fan of the extra complexity it would entail 😕 @merceyz why not as immutable pattern? This would integrate with the existing mechanism, without need to implement it again in `persistManifest`.

**Comment by isker** [comment]
@arcanis @merceyz would you then consider a configuration flag to disable this manifest reformatting?  We already use [prettier](https://github.com/prettier/prettier) to format all of our code; as it stands for users in our situation, yarn's formatting is a liability.

They even conflict in some situations, e.g. prettier would produce this
```json
  "workspaces": ["foo/*/*", "bar/*"],
```
whereas yarn produces this
```json
  "workspaces": [
    "foo/*/*",
    "bar/*"
  ],
```
in our root package.json.  Everything else being equal, we'd prefer to have only the tool whose entire purpose is the imposition of aesthetics imposing its aesthetics upon us 😃.

**Comment by arcanis** [comment]
I'd tend toward no, because Yarn isn't "formatting" thing intentionally, so it's difficult to stop doing it. We serialize the file via `JSON.stringify`, and that's where the formatting gets applied. We can't just stop generating the files (after all that's how we detect whether the file would change), so necessarily the manifests needs to have the format `JSON.stringify` expects. The only other option would be to detect Prettier's presence and automatically run it, but that's out of scope for Yarn.

My recommendation would probably be to just ignore the `package.json` files from Prettier.

**Comment by isker** [comment]
@arcanis can you explain why yarn needs to write these files at all?  At face value it seems like "intentional formatting"; is it serving any other purpose?

**Comment by arcanis** [comment]
Yes; mutations that you make prior to running the install are persisted on disk. This is for example how `yarn add` and `yarn remove` work (we don't "parse, write, re-parse, install": we "parse, mutate, install, write").
